blockly-ws-search/src/WorkspaceSearch.js

```diff
@@ -0,0 +1,660 @@
+
+    // Create the button
+    const btn = document.createElement('button');
+    Blockly.utils.dom.addClass(btn, className);
+    btn.append(textSpan);
+    return btn;
+  }
+
+  /**
+   * Add event listener for clicking and keydown on the given button.
+   * @param {!HTMLButtonElement} btn The button to add the event listener to.
+   * @param {!Function} onClickFn The function to call when the user clicks on 
+   *     or hits enter on the button.
+   * @private
+   */
+  addBtnListener_(btn, onClickFn) {
+    btn.addEventListener('click', onClickFn);
+    // TODO: Review Blockly's key handling to see if there is a way to avoid
+    //  needing to call stopPropogation().
+    btn.addEventListener('keydown', e => {
+      if (e.key === "Enter") {
+        onClickFn(e);
+        e.preventDefault();  
+      } else if (e.key === "Escape") {
+        this.close();
+      }
+      e.stopPropagation();
+    });
+  }
+
+  /**
+   * Positions the search bar based on where the workspace's toolbox is.
+   */
+  positionSearchBar() {
+    // TODO: Handle positioning search bar when window is resized.
+    const metrics = this.workspace_.getMetrics();
+    if (this.workspace_.RTL) {
+      this.htmlDiv_.style.left = metrics.absoluteLeft + 'px';
+    } else {
+      if (metrics.toolboxPosition == Blockly.TOOLBOX_AT_RIGHT) {
+        this.htmlDiv_.style.right = metrics.toolboxWidth + 'px';
+      } else {
+        this.htmlDiv_.style.right = '0';
+      }
+    }
+    this.htmlDiv_.style.top = metrics.absoluteTop + 'px';
+  }
+
+  /**
+   * Handles input value change in search bar.
+   * @private
+   */
+  onInput_() {
+    if (this.searchOnInput) {
+      const inputValue = this.inputElement_.value;
+      if (inputValue !== this.searchText_) {
+        this.setSearchText_(inputValue);
+        this.searchAndHighlight(this.preserveSelected);
+      }
+    }
+  }
+
+  /**
+   * Handles a key down for the search bar.
+   * @param {KeyboardEvent} e The key down event.
+   * @private
+   */
+  onKeyDown_(e) {
+    if (e.key === 'Escape') {
+      this.close();
+    } else if (e.key === 'Enter') {
+      if (this.searchOnInput) {
+        this.next();
+      } else {
+        this.setSearchText_(this.inputElement_.value);
+        this.searchAndHighlight(this.preserveSelected);
+      }
+    }
+  }
+
+  /**
+   * Opens the search bar when Control F or Command F are used on the workspace.
+   * @param {KeyboardEvent} e The key down event.
+   * @private
+   */
+  onWorkspaceKeyDown_(e) {
+    // TODO: Look into handling keyboard shortcuts on workspace in Blockly.
+    if ((e.ctrlKey || e.metaKey) && e.key === "f") {
+      this.open();
+      e.preventDefault();
+    }
+  }
+
+  /**
+   * Selects the previous block.
+   */
+  previous() {
+    if (!this.blocks_.length) {
+      return;
+    }
+    this.setCurrentBlock_(this.currentBlockIndex_ - 1);
+  }
+
+  /**
+   * Selects the next block.
+   */
+  next() {
+    if (!this.blocks_.length) {
```

Comment:
No need for this, you're not doing any real validation here. Validation is already happening in setCurrentBlock_.